Silabs USART split series 0/1 and series 2 board driver

Martinhoff-maker opened the pull request:
The goal of this PR is to split USART driver for series 0/1 and series 2 Silabs boards. It allows better maintainability (because Silabs series 2  boards support pin-ctrl and clock-ctrl)

asmellby commented on dts/bindings/serial/silabs,usart.yaml:
```diff
@@ -0,0 +1,12 @@
+description: Silabs USART
+
+compatible: "silabs,usart"
```
Should this be `silabs,usart-uart` to align with https://github.com/zephyrproject-rtos/zephyr/pull/83232 ?

Martinhoff-maker replied:
yes you're right, I will fix it.

asmellby commented on drivers/serial/Kconfig.silabs_usart:
```diff
@@ -0,0 +1,21 @@
+	select SERIAL_HAS_DRIVER
+	select SERIAL_SUPPORT_INTERRUPT
+	select SOC_GECKO_USART
+	# ASYNC implementation requires a DMA controller
```
Async implementation doesn't seem to exist yet? I assume it will come in a followup PR?

Martinhoff-maker replied:
Yes, it will come in the followup PR

asmellby commented on drivers/serial/Kconfig.silabs_usart:
```diff
@@ -0,0 +1,21 @@
+# Silabs USART
```
nit: Unnecessary comment

jhedberg commented:
@Martinhoff-maker looks pretty good, but please rebase to fix the conflicts with the latest main branch.

Martinhoff-maker commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -261,6 +261,9 @@ Serial
 ======
 
 * Renamed the ``compatible`` from ``nxp,kinetis-lpuart`` to :dtcompatible:`nxp,lpuart`.
+* Renamed the ``compatible`` from ``silabs,gecko-usart`` to :dtcompatible:`silabs,usart-uart`.
```
Good question. I don't really know how to handle this case in the migration guide. that's why I wrote the next line. If I removed it, do you think the next line is clear enough to understand the change ?

yishai1999 replied:
Ya for sure. There was only one action done in this PR and it's splitting the driver into two.

Martinhoff-maker committed:
driver: serial: split silabs series 2 and series 0/1 boards usart driver

Split the USART driver into separate implementations for Silabs Series 2
and Series 0/1 boards. This change improves maintainability (especially
with the support of pin-ctrl and clock-ctrl on series 2 boards).

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

Martinhoff-maker commented:
Btw, I split the driver before @yishai1999 added to runtime configure option. That's why the feature is not yet supported in the driver for series 2 boards. This feature will come next with my followup PR that add DMA support for Usart.

jerome-pouiller commented on drivers/serial/uart_silabs_usart.c:
```diff
@@ -0,0 +1,352 @@
+	const struct uart_silabs_config *config = dev->config;
+	int num_tx = 0U;
+
+	while ((len - num_tx > 0) && (config->base->STATUS & USART_STATUS_TXBL)) {
```
Average developer would write `num_tx < len` rather than `(len - num_tx > 0)`.

jerome-pouiller commented on drivers/serial/uart_silabs_usart.c:
```diff
@@ -0,0 +1,352 @@
+static int uart_silabs_fifo_fill(const struct device *dev, const uint8_t *tx_data, int len)
+{
+	const struct uart_silabs_config *config = dev->config;
+	int num_tx = 0U;
```
Indexes are usually named `i` (rather than `num_tx`).

jerome-pouiller commented on drivers/serial/uart_silabs_usart.c:
```diff
@@ -0,0 +1,352 @@
+		rx_data[num_rx++] = (uint8_t)config->base->RXDATA;
+	}
+
+	return num_rx;
```
Same comments than `uart_silabs_fifo_fill()`

jerome-pouiller commented on drivers/serial/uart_silabs_usart.c:
```diff
@@ -0,0 +1,352 @@
+
+	USART_IntClear(config->base, USART_IF_TXC);
+
+	return (flags & USART_IF_TXC) != 0U;
```
Tips: you can write `!!(flags & USART_IF_TXC)`

jerome-pouiller commented on drivers/serial/uart_silabs_usart.c:
```diff
@@ -0,0 +1,352 @@
+	const struct uart_silabs_config *config = dev->config;
+	uint32_t mask = USART_IEN_TXBL | USART_IEN_TXC;
+
+	USART_IntEnable(config->base, mask);
```
`uart_silabs_irq_err_enable()` do not introduce `mask` variable. I suggest to also inline `mask` in `uart_silabs_irq_[tr]x_*()`.

Martinhoff-maker replied:
done

jerome-pouiller commented on drivers/serial/uart_silabs_usart.c:
```diff
@@ -0,0 +1,352 @@
+ *
+ * @param dev UART device to be initialized
+ * @return int 0
+ */
```
Probably this comment does not bring any value.

jerome-pouiller commented on drivers/serial/uart_silabs_usart.c:
```diff
@@ -0,0 +1,352 @@
+
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+#ifdef USART_STATUS_TXIDLE
```
`USART_STATUS_TXIDLE` is always defined in `simplicity_sdk`:
```
$ find simplicity_sdk -name '*_usart.h' | xargs  git grep -L 'define USART_STATUS_TXIDLE\b'
$

jerome-pouiller commented on drivers/serial/uart_silabs_usart.c:
```diff
@@ -0,0 +1,352 @@
+		.flow_ctrl = DT_INST_PROP(idx, hw_flow_control) ? UART_CFG_FLOW_CTRL_RTS_CTS       \
+								: UART_CFG_FLOW_CTRL_NONE,         \
+		SILABS_USART_IRQ_HANDLER_FUNC(idx)                                                 \
+		};                                                                                 \
```
Do not indnent this brace.

jerome-pouiller commented on drivers/serial/uart_silabs_usart.c:
```diff
@@ -0,0 +1,352 @@
+		.clock_cfg = SILABS_DT_INST_CLOCK_CFG(idx),                                        \
+		.baud_rate = DT_INST_PROP(idx, current_speed),                                     \
+		.flow_ctrl = DT_INST_PROP(idx, hw_flow_control) ? UART_CFG_FLOW_CTRL_RTS_CTS       \
+								: UART_CFG_FLOW_CTRL_NONE,         \
```
I suggest to either declare `flow_ctrl` as a `bool` or replace `UART_CFG_FLOW_CTRL_*` by `usartHwFlowControlCtsAndRts` and `usartHwFlowControlNone`.

Martinhoff-maker replied:
replace with usartHwFlowControlCtsAndRts and usartHwFlowControlNone. It will change with the followup PR which introduce runtime configuration.

Martinhoff-maker committed:
dts: arm: change usart binding of silabs series 2 boards

Make the Silabs series 2 boards use the new USART driver
"silabs,usart-uart".

Signed-off-by: Martin Hoff <martin.hoff@silabs.com>

yishai1999 commented:
> Btw, I split the driver before @yishai1999 added the runtime configure option. That's why the feature is not yet supported in the driver for series 2 boards. This feature will come next with my followup PR that add DMA support for Usart.

This would mean you are removing an existing feature. I think this should be added in this PR and not in a separate one.

yishai1999 commented on doc/releases/migration-guide-4.1.rst:
```diff
@@ -275,6 +275,8 @@ Serial
 ======
 
 * Renamed the ``compatible`` from ``nxp,kinetis-lpuart`` to :dtcompatible:`nxp,lpuart`.
+* Silabs Usart driver have been splited for Series 2 :dtcompatible:`silabs,usart-uart`
```
nit: `has been split for...`

jhedberg commented:
@dcpleung please take another look, thanks!

kartben commented:
@Martinhoff-maker where did `uart_gecko_pm_action` go? Driver fails to compile when PM_DEVICE is enabled. Not familiar with the driver at all but it might be that you also wanted to remove the two `PM_DEVICE_DT_INST_DEFINE(idx, uart_gecko_pm_action);` lines

https://github.com/zephyrproject-rtos/zephyr/actions/runs/12970883619/job/36176680696#step:13:4237

@jerome-pouiller @jhedberg

Martinhoff-maker commented:
@kartben `PM_DEVICE_DT_INST_DEFINE(idx, uart_gecko_pm_action)` needs to be here but I made a mistake (delating `uart_gecko_pm_action` function) when deleting clock control part in this driver. I will create a Hotfix PR. Sorry for that.
